text/0091-dht-record-creation-time.md

```diff
@@ -0,0 +1,105 @@
+
+After this RFC, nodes are extended to decide to keep the new record and propagate the new record to nodes that have the old record stored, so in the end all the nodes will converge faster to the new record(in the order of minutes, not 36h)
+
+Implementation of the rfc: https://github.com/paritytech/polkadot-sdk/pull/3786.
+
+Current issue without this enhacement: https://github.com/paritytech/polkadot-sdk/issues/3673 
+
+## Stakeholders
+
+Polkadot node developers.
+
+## Explanation
+
+This RFC heavily relies on the functionalities of the Kademlia DHT already in use by Polkadot.
+You can find a link to the specification [here](https://github.com/libp2p/specs/tree/master/kad-dht).
+
+In a nutshell, on a specific node the current authority-discovery protocol publishes Kademila DHT records at startup and periodically. The records contain the full address of the node for each authorithy key it owns. The node tries also to find the full address of all authorities in the network by querying the DHT and picking up the first record it finds for each of the authority id it found on chain.
+
+The authority discovery DHT records use the protobuf protocol and the current format is specified [here](https://github.com/paritytech/polkadot-sdk/blob/313fe0f9a277f27a4228634f0fb15a1c3fa21271/substrate/client/authority-discovery/src/worker/schema/dht-v2.proto#L4). This RFC proposese extending the schema in a backwards compatible manner by adding a new optional `creation_time` field to `AuthorityRecord` and nodes can use this information to determine which of the record is newer.
+
+Diff of `dht-v3.proto` vs `dht-v2.proto`
+
+```
+@@ -1,10 +1,10 @@
+ syntax = "proto3";
+
+-package authority_discovery_v2;
++package authority_discovery_v3;
+
+ // First we need to serialize the addresses in order to be able to sign them.
+ message AuthorityRecord {
+ 	repeated bytes addresses = 1;
++	// Time since UNIX_EPOCH in nanoseconds, scale encoded
++	TimestampInfo creation_time = 2;
+ }
+
+ message PeerSignature {
+@@ -13,11 +15,17 @@
+ 	bytes public_key = 2;
+ }
+
++// Information regarding the creation data of the record
++message TimestampInfo {
++       // Time since UNIX_EPOCH in nanoseconds, scale encoded
++       bytes timestamp = 1;
++}
++
+```
+
+Each time a node wants to resolve an authorithy ID it will issue a query with a certain redundancy factor, and from all the results it receives it will decide to pick only the newest record. Additionally, the nodes that answer with old records will be updated with the newer record.
```

Comment:
> Additionally, the nodes that answer with old records will be updated with the newer record.

So, the node that send the DHT request will inform these nodes?

Strictly speaking, I don't think that this needs to be part of the RFC, because even a node not doing this would behave well.

Reply:
Yes, that's correct, I was just trying to explain how a well behaving validator should/could use this field, I can explicitly state that this part is optional or remove it.